File: JsTestDriver/src/com/google/jstestdriver/Response.java
/*
 * Copyright 2009 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.jstestdriver;

import java.lang.reflect.Type;
import java.util.Collection;

import com.google.gson.reflect.TypeToken;


/**
 * The Response from the browser by way of the server.
 * @author jeremiele@google.com (Jeremie Lenfant-Engelmann)
 */
// TODO(corysmith): move to the protocol package
public class Response {
  
  public static enum ResponseType {
    FILE_LOAD_RESULT(LoadedFiles.class),
    REGISTER_RESULT(null),
    TEST_RESULT(new TypeToken<Collection<TestResult>>() {}.getType()),
    TEST_QUERY_RESULT(null),
    RESET_RESULT(null),
    BROWSER_PANIC(BrowserPanic.class),
    UNKNOWN(null),
    BROWSER_READY(null),
    COMMAND_RESULT(null),
    LOG(null);

    public final Type type;

    ResponseType(Type type) {
      this.type = type;
    }
  }

  private String type;
  private String response = "";
  private BrowserInfo browser = new BrowserInfo();
  private String error = "";
  private long executionTime = 0L;

  public String getResponse() {
    return response;
  }

  public void setResponse(String response) {
    this.response = response;
  }

  public BrowserInfo getBrowser() {
    return browser;
  }

  public void setBrowser(BrowserInfo browser) {
    this.browser = browser;
  }

  public String getError() {
    return error;
  }

  public void setError(String error) {
    this.error = error;
  }

  public long getExecutionTime() {
    return executionTime;
  }

  public void setExecutionTime(long executionTime) {
    this.executionTime = executionTime;
  }
  
  /** Sets the string representation of the type. */
  public void setType(String type) {
    this.type = type;
  }
  
  /** Gets the string representation of the type. */
  public ResponseType getResponseType() {
    if (type == null) {
      return ResponseType.UNKNOWN;
    }
    return ResponseType.valueOf(type);
  }
  
  /** The type for Gson use to deserialize the response. */
  public Type getGsonType() {
    return ResponseType.valueOf(type).type;
  }

  @Override
  public String toString() {
    return "Response (\nbrowser=[" + browser + "], \nerror=[" + error + "], \nexecutionTime=[" + executionTime
        + "], \nresponse=[" + response + "], \ntype=[" + type + "])";
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((browser == null) ? 0 : browser.hashCode());
    result = prime * result + ((error == null) ? 0 : error.hashCode());
    result = prime * result + (int) (executionTime ^ (executionTime >>> 32));
    result = prime * result + ((response == null) ? 0 : response.hashCode());
    result = prime * result + ((type == null) ? 0 : type.hashCode());
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (!(obj instanceof Response)) return false;
    Response other = (Response) obj;
    if (browser == null) {
      if (other.browser != null) return false;
    } else if (!browser.equals(other.browser)) return false;
    if (error == null) {
      if (other.error != null) return false;
    } else if (!error.equals(other.error)) return false;
    if (executionTime != other.executionTime) return false;
    if (response == null) {
      if (other.response != null) return false;
    } else if (!response.equals(other.response)) return false;
    if (type == null) {
      if (other.type != null) return false;
    } else if (!type.equals(other.type)) return false;
    return true;
  }
}

File: JsTestDriver/src/com/google/jstestdriver/javascript/Response.js
/*
 * Copyright 2010 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
/**
 * A series of constants to correlate to the type of responses between the
 * browser and the server.
 * 
 * See com.google.jstestdriver.Response.GSON_TYPES.
 * @author corbinrsmith@gmail.com (Cory Smith)
 */
jstestdriver.RESPONSE_TYPES = {
  FILE_LOAD_RESULT: 'FILE_LOAD_RESULT',
  REGISTER_RESULT: 'REGISTER_RESULT',
  TEST_RESULT: 'TEST_RESULT',
  TEST_QUERY_RESULT: 'TEST_QUERY_RESULT',
  RESET_RESULT: 'RESET_RESULT',
  COMMAND_RESULT: 'COMMAND_RESULT',
  BROWSER_READY: 'BROWSER_READY',
  BROWSER_PANIC: 'BROWSER_PANIC',
  LOG: 'LOG',
};


/**
 * Contains the state of a response.
 * This is the javascript twin to com.google.jstestdriver.Response.
 * 
 * @param {jstestdriver.RESPONSE_TYPES} type The type of the response.
 * @param {String} response The serialized contents of the response.
 * @param {jstestdriver.BrowserInfo} browser The browser information. 
 * @param {Boolean} start Is this the first response from the browser.
 */
jstestdriver.Response = function(type, response, browser, start) {
  this.type = type;
  this.response = response;
  this.browser = browser;
  if (start) {
    this.start = true;
  }
};



/**
 * @param {String} done Indicates if this is the last streamed message.
 * @param {jstestdriver.Response} response The response.
 */
jstestdriver.CommandResponse = function (done, response) {
  this.done = done;
  this.response = response;
};



/**
 * Represents the information about the browser.
 * @param {Number} id The unique id of this browser.
 */
jstestdriver.BrowserInfo = function(id) {
  this.id = id;
};


/**
 * Coincides with java class com.google.jstestdriver.protocol.BrowserLog.
 * @param {Number} level 1-5, the level of the message.
 */
jstestdriver.BrowserLog = function(level, source, message, browser) {
  this.level = level;
  this.source = source;
  this.message = message;
  this.browser = browser;
}
